File: generate-icons.js
// Generate PWA icon PNGs for Sophia Oracle
// Run: node generate-icons.js
// Creates icons/icon-192.png, icons/icon-512.png, etc.

const { createCanvas } = require('canvas');
const fs = require('fs');
const path = require('path');

const ICONS_DIR = path.join(__dirname, 'icons');
if (!fs.existsSync(ICONS_DIR)) fs.mkdirSync(ICONS_DIR);

function drawIcon(size, maskable = false) {
    const c = createCanvas(size, size);
    const ctx = c.getContext('2d');
    const s = size / 64; // scale from 64px base

    // Background
    const bgR = size * 0.22;
    ctx.fillStyle = '#0a0a0f';
    if (maskable) {
        // Maskable icons: fill entire rect, no rounded corners
        ctx.fillRect(0, 0, size, size);
    } else {
        ctx.beginPath();
        ctx.roundRect(0, 0, size, size, bgR);
        ctx.fill();
    }

    // Subtle radial glow behind crystal ball
    const cx = size / 2, cy = size * 0.42, r = size * 0.25;
    const glowGrad = ctx.createRadialGradient(cx, cy, r * 0.3, cx, cy, r * 2);
    glowGrad.addColorStop(0, 'rgba(123, 80, 212, 0.25)');
    glowGrad.addColorStop(0.5, 'rgba(255, 0, 85, 0.08)');
    glowGrad.addColorStop(1, 'rgba(0, 0, 0, 0)');
    ctx.fillStyle = glowGrad;
    ctx.fillRect(0, 0, size, size);

    // Crystal ball — main sphere with gradient
    const ballGrad = ctx.createRadialGradient(cx - r * 0.2, cy - r * 0.3, r * 0.1, cx, cy, r);
    ballGrad.addColorStop(0, '#c4a0ff');
    ballGrad.addColorStop(0.3, '#9B6FE8');
    ballGrad.addColorStop(0.6, '#7B50D4');
    ballGrad.addColorStop(1, '#2a1050');
    ctx.beginPath();
    ctx.arc(cx, cy, r, 0, Math.PI * 2);
    ctx.fillStyle = ballGrad;
    ctx.fill();

    // Glass highlight (top-left shine)
    const shineGrad = ctx.createRadialGradient(cx - r * 0.35, cy - r * 0.35, r * 0.05, cx - r * 0.2, cy - r * 0.2, r * 0.5);
    shineGrad.addColorStop(0, 'rgba(255, 255, 255, 0.55)');
    shineGrad.addColorStop(0.5, 'rgba(255, 255, 255, 0.15)');
    shineGrad.addColorStop(1, 'rgba(255, 255, 255, 0)');
    ctx.beginPath();
    ctx.arc(cx, cy, r * 0.95, 0, Math.PI * 2);
    ctx.fillStyle = shineGrad;
    ctx.fill();

    // Inner glow — mystical energy
    const innerGlow = ctx.createRadialGradient(cx, cy + r * 0.2, r * 0.05, cx, cy, r * 0.7);
    innerGlow.addColorStop(0, 'rgba(255, 0, 85, 0.3)');
    innerGlow.addColorStop(0.5, 'rgba(196, 160, 255, 0.1)');
    innerGlow.addColorStop(1, 'rgba(0, 0, 0, 0)');
    ctx.beginPath();
    ctx.arc(cx, cy, r * 0.8, 0, Math.PI * 2);
    ctx.fillStyle = innerGlow;
    ctx.fill();

    // Sphere rim — subtle pink glow
    ctx.beginPath();
    ctx.arc(cx, cy, r, 0, Math.PI * 2);
    ctx.strokeStyle = 'rgba(255, 0, 85, 0.4)';
    ctx.lineWidth = s * 1.5;
    ctx.stroke();

    // Outer ethereal ring
    ctx.beginPath();
    ctx.arc(cx, cy, r + s * 2, 0, Math.PI * 2);
    ctx.strokeStyle = 'rgba(196, 160, 255, 0.15)';
    ctx.lineWidth = s * 1;
    ctx.stroke();

    // Base / stand
    const baseY = cy + r + s * 2;
    ctx.beginPath();
    ctx.moveTo(cx - r * 0.8, baseY);
    ctx.quadraticCurveTo(cx - r * 0.9, baseY + s * 5, cx - r * 0.95, baseY + s * 8);
    ctx.lineTo(cx + r * 0.95, baseY + s * 8);
    ctx.quadraticCurveTo(cx + r * 0.9, baseY + s * 5, cx + r * 0.8, baseY);
    ctx.closePath();
    const baseGrad = ctx.createLinearGradient(cx, baseY, cx, baseY + s * 8);
    baseGrad.addColorStop(0, 'rgba(255, 0, 85, 0.35)');
    baseGrad.addColorStop(1, 'rgba(255, 0, 85, 0.1)');
    ctx.fillStyle = baseGrad;
    ctx.fill();
    ctx.strokeStyle = 'rgba(255, 0, 85, 0.3)';
    ctx.lineWidth = s * 0.5;
    ctx.stroke();

    // Small stars / sparkles
    const stars = [
        { x: cx - r * 1.4, y: cy - r * 0.8, s: 2 },
        { x: cx + r * 1.3, y: cy - r * 0.6, s: 1.5 },
        { x: cx - r * 1.1, y: cy + r * 0.4, s: 1 },
        { x: cx + r * 1.5, y: cy + r * 0.2, s: 1.8 },
        { x: cx + r * 0.5, y: cy - r * 1.5, s: 1.2 },
    ];
    stars.forEach(star => {
        ctx.beginPath();
        ctx.arc(star.x, star.y, s * star.s, 0, Math.PI * 2);
        ctx.fillStyle = `rgba(196, 160, 255, ${0.3 + Math.random() * 0.3})`;
        ctx.fill();
    });

    return c;
}

// Generate all icon variants
const configs = [
    { size: 192, file: 'icon-192.png', maskable: false },
    { size: 512, file: 'icon-512.png', maskable: false },
    { size: 192, file: 'icon-maskable-192.png', maskable: true },
    { size: 512, file: 'icon-maskable-512.png', maskable: true },
];

configs.forEach(({ size, file, maskable }) => {
    const canvas = drawIcon(size, maskable);
    const buffer = canvas.toBuffer('image/png');
    const outPath = path.join(ICONS_DIR, file);
    fs.writeFileSync(outPath, buffer);
    console.log(`✅ Generated ${outPath} (${(buffer.length / 1024).toFixed(1)} KB)`);
});

console.log('\n📲 All PWA icons generated! Commit the icons/ folder.');

File: generate-icons.py
"""Generate PWA icon PNGs for Sophia Oracle. Run: python generate-icons.py"""
import os, math, random
from PIL import Image, ImageDraw

ICONS_DIR = os.path.join(os.path.dirname(__file__), "icons")
os.makedirs(ICONS_DIR, exist_ok=True)


def soft_circle(draw, cx, cy, radius, color):
    """Draw a soft-edged circle with falloff."""
    r, g, b, peak_a = color
    for i in range(radius, 0, -1):
        t = (radius - i) / radius  # 0 at edge, 1 at center
        a = int(peak_a * t * t)    # quadratic falloff for soft edges
        if a < 1: continue
        draw.ellipse([cx - i, cy - i, cx + i, cy + i], fill=(r, g, b, a))


def draw_icon(size, maskable=False):
    # Work on a 2x canvas for antialiasing, then downscale
    ss = 2  # supersample factor
    sz = size * ss
    img = Image.new("RGBA", (sz, sz), (0, 0, 0, 0))
    draw = ImageDraw.Draw(img)
    s = sz / 64
    cx, cy = sz // 2, int(sz * 0.40)
    r = int(sz * 0.23)

    # ── Solid dark background ──
    bg = (10, 8, 18, 255)
    if maskable:
        draw.rectangle([0, 0, sz, sz], fill=bg)
    else:
        draw.rounded_rectangle([0, 0, sz - 1, sz - 1],
                               radius=int(sz * 0.20), fill=bg)

    # ── Minimal ambient glow — barely visible ──
    soft_circle(draw, cx, cy, int(r * 1.4), (40, 20, 80, 10))

    # ── Crystal ball — dark-to-medium purple sphere ──
    # Outer dark ring
    draw.ellipse([cx - r, cy - r, cx + r, cy + r], fill=(20, 8, 45, 255))
    # Inner gradient: dark edge to slightly brighter center
    for i in range(r - 1, 0, -1):
        t = 1.0 - (i / r)  # 0 at edge, 1 at center
        t2 = t * t          # ease in
        red = int(20 + 100 * t2)
        grn = int(8 + 60 * t2)
        blu = int(45 + 140 * t2)
        draw.ellipse([cx - i, cy - i, cx + i, cy + i],
                     fill=(red, grn, blu, 255))

    # ── Inner mystical glow — subtle magenta/pink deep inside ──
    soft_circle(draw, cx, cy + int(r * 0.1), int(r * 0.45),
                (200, 40, 120, 50))

    # ── Glass highlight — small, crisp, top-left ──
    hx, hy = cx - int(r * 0.30), cy - int(r * 0.30)
    soft_circle(draw, hx, hy, int(r * 0.16), (255, 255, 255, 220))

    # ── Tiny catch light bottom-right ──
    soft_circle(draw, cx + int(r * 0.22), cy + int(r * 0.30),
                int(r * 0.06), (255, 255, 255, 80))

    # ── Sphere rim — very subtle magenta edge ──
    lw = max(1, int(s * 0.5))
    draw.ellipse([cx - r, cy - r, cx + r, cy + r],
                 outline=(200, 0, 70, 40), width=lw)

    # ── Very faint outer aura ──
    aura_r = r + int(s * 1.5)
    draw.ellipse([cx - aura_r, cy - aura_r, cx + aura_r, cy + aura_r],
                 outline=(120, 80, 200, 10), width=max(1, int(s * 0.3)))

    # ── Base / stand ──
    base_top = cy + r + int(s * 2.5)
    base_h = int(s * 6)
    tw = int(r * 0.45)
    bw = int(r * 0.7)
    for row in range(base_h):
        t = row / base_h
        w = int(tw + (bw - tw) * t)
        y = base_top + row
        a = int(35 + 25 * (1 - t))
        draw.line([(cx - w, y), (cx + w, y)], fill=(180, 20, 70, a), width=1)
    draw.polygon([
        (cx - tw, base_top), (cx + tw, base_top),
        (cx + bw, base_top + base_h), (cx - bw, base_top + base_h),
    ], outline=(255, 0, 85, 25))

    # ── Tiny stars — scattered, dim ──
    random.seed(42)
    for _ in range(10):
        angle = random.uniform(0, math.pi * 2)
        dist = random.uniform(r * 1.3, r * 2.4)
        sx = int(cx + math.cos(angle) * dist)
        sy = int(cy + math.sin(angle) * dist * 0.9)
        margin = int(s * 5)
        if sx < margin or sx > sz - margin: continue
        if sy < margin or sy > sz - margin: continue
        sr = max(1, int(s * random.uniform(0.4, 1.2)))
        brightness = random.randint(130, 200)
        soft_circle(draw, sx, sy, sr,
                    (brightness, brightness - 40, 255, 120))

    # ── Cross sparkle on highlight ──
    sp_len = int(s * 2.5)
    for i in range(sp_len, 0, -1):
        a = int(60 * (1 - i / sp_len))
        draw.line([(hx - i, hy), (hx + i, hy)], fill=(255, 255, 255, a), width=1)
        draw.line([(hx, hy - i), (hx, hy + i)], fill=(255, 255, 255, a), width=1)

    # Downscale with antialiasing
    img = img.resize((size, size), Image.LANCZOS)
    return img


configs = [
    (192, "icon-192.png", False),
    (512, "icon-512.png", False),
    (192, "icon-maskable-192.png", True),
    (512, "icon-maskable-512.png", True),
]

for size, filename, maskable in configs:
    icon = draw_icon(size, maskable)
    path = os.path.join(ICONS_DIR, filename)
    icon.save(path, "PNG", optimize=True)
    fsize = os.path.getsize(path)
    print(f"  Created {path} ({fsize // 1024} KB)")

print("\nAll PWA icons generated!")
